main/lsp/requests/hover.cc

```diff
@@ -24,11 +24,38 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
     return contents;
 }
 
-unique_ptr<MarkupContent> formatRubyCode(MarkupKind markupKind, string str) {
-    if (markupKind == MarkupKind::Markdown && str.length() > 0) {
-        str = fmt::format("```ruby\n{}\n```", str);
+unique_ptr<MarkupContent> formatRubyCode(MarkupKind markupKind, string_view sigString,
+                                         optional<string_view> optionalDocString) {
+    string_view docString = optionalDocString.value_or("");
+    // Get rid of at most 1 trailing newline
+    if (docString.length() > 0 && docString.back() == '\n') {
+        docString.remove_suffix(1);
     }
-    return make_unique<MarkupContent>(markupKind, move(str));
+    if (sigString.length() > 0 && sigString.back() == '\n') {
+        sigString.remove_suffix(1);
+    }
+
+    string content = "";
```

Comment:
If `findDocumentation` returns `nullopt` if the documentation string is zero-length after trimming, then we can just do something like:

```
if (optionalDocString.has_value()) {
  return absl::StrCat(optionalDocString.value(), "\n", formattedSigString);
} else {
  return formattedSigString;
}
```

Reply:
This looks good, but we still need to call make_unique<MarkupContent> before return. It would look something like 
```cpp
    string content; 
    if(docString.has_value()) {
        content = absl::StrCat(docString.value(), formattedSigString);
    } else {
        content = formattedSigString;
    }
    return make_unique<MarkupContent>(markupKind, move(content));
```
With the invariant that non-empty docs always have a newline, always appending might be easier to read I think

Reply:
> With the invariant that non-empty docs always have a newline,

Why is this an invariant?

Reply:
> With the invariant that non-empty docs always have a newline

I don't see the issue? `if (optionalDocString.has_value())`, then it contains a non-empty docstring. No need for a conditional to see if it's zero-length.

I think it's easier to read if we see all of the elements coming together in a single location. The fewer append operations, the better. I also think this method is the right place to append a newline after the docs, not `findDocumentation`.

Reply:
Why not:

```c++
string formattedTypeString;
/** code that sets formattedTypeString **/

return make_unique<MarkupContent>(markupKind,  absl::StrCat(docString.value_or(""), formattedTypeString));
```

That removes a bunch of LOC and `StrAppend` operations.

Reply:
5562c3adf